File: GUI/Publish/JScript/Application.js
/*
=============================================
Author      : <ยุทธภูมิ ตวันนา>
Create date : <๒๖/๐๗/๒๕๖๑>
Modify date : <๑๙/๐๙/๒๕๖๒>
Description : <รวมรวบฟังก์ชั่นใช้งานสำหรับใบสมัคร>
=============================================
*/

(function () {
	"use strict";

    angular.module("applicationMod", [
	    "utilMod",
	    "appMod",
	    "dictMod",
	    "pageRouteMod"
    ])
    
    .service("applicationServ", function (
        $rootScope,
        $timeout,
        $location,
        $q,
        utilServ,
        appServ,
        dictServ,
        pageRouteServ
    ) {
        var self = this;

        self.isSubmitStatus = false;

        self.getDataSource = function (param) {
		    param.dataSource = (param.dataSource === undefined || param.dataSource === "" ? [] : param.dataSource);

		    var deferred = $q.defer();
		    var cookie = utilServ.getCookie({ cookieName: appServ.cookieName });

		    appServ.getCookieValue(["userId", "verifyCode", "applicationId", "applicationStatus"]);

            if (param.dataSource.length === 0) {
                appServ.getListApplication({
				    action: "get",
				    params: [
					    "",
					    ("package=" + appServ.setPackage([cookie.key, appServ.cookieInfo.applicationId, appServ.cookieInfo.userId, appServ.cookieInfo.verifyCode]))
				    ].join("&")
                }).then(function (result) {
                    var dt = [];

                    if (result.length > 0) {
                        for (var i = 0; i < result.length; i++) {
                            var dr = result[i];

                            dt.push({
                                id: (dr.id ? dr.id : ""),
                                UserId: (dr.sexUserId ? dr.sexUserId : ""),
                                applicationStatus: (dr.applicationStatus ? dr.applicationStatus : ""),
                                applicationStatusName: (dr.applicationStatusName ? dr.applicationStatusName : ""),
                                titlePrefixId: (dr.perTitlePrefixId ? dr.perTitlePrefixId : ""),
                                titlePrefixFullName: {
								    TH: "",
                                    EN: (dr.titlePrefixFullNameEN ? dr.titlePrefixFullNameEN : "")
                                },
                                titlePrefixInitials: {
                                    TH: "",
                                    EN: (dr.titlePrefixInitialsEN ? dr.titlePrefixInitialsEN : "")
                                },
                                firstName: (dr.firstName ? dr.firstName : ""),
                                middleName: (dr.firstName ? dr.middleName : ""),
                                lastName: (dr.lastName ? dr.lastName : ""),
                                fullName: (dr.fullName ? dr.fullName : ""),
                                nationalityId: (dr.perNationalityId ? dr.perNationalityId : ""),
                                nationalityName: {
                                    TH: "",
                                    EN: (dr.nationalityNameEN ? dr.nationalityNameEN : "")
                                },
                                isoNationalityName2Letter: (dr.isoNationalityName2Letter ? dr.isoNationalityName2Letter : ""),
                                isoNationalityName3Letter: (dr.isoNationalityName3Letter ? dr.isoNationalityName3Letter : ""),
                                genderId: (dr.perGenderId ? dr.perGenderId : ""),
                                genderFullName: {
								    TH: "",
                                    EN: (dr.genderFullNameEN ? dr.genderFullNameEN : "")
                                },
                                genderInitials: {
                                    TH: "",
                                    EN: (dr.genderInitialsEN ? dr.genderInitialsEN : "")
                                },
                                email: (dr.email ? dr.email : ""),
                                studentCategoryId: (dr.sexStudentCategoryId ? dr.sexStudentCategoryId : ""),
                                studentCategoryName: {
                                    TH: "",
                                    EN: (dr.studentCategoryNameEN ? dr.studentCategoryNameEN : "")
                                },
                                studentCategorySpecify: (dr.studentCategorySpecify ? dr.studentCategorySpecify : ""),
                                studentCategoryOther: (dr.studentCategoryOther ? dr.studentCategoryOther : ""),
                                countryHomeUniversityId: (dr.countryHomeUniversityId ? dr.countryHomeUniversityId : ""),
                                countryHomeUniversityName: {
                                    TH: "",
                                    EN: (dr.countryHomeUniversityNameEN ? dr.countryHomeUniversityNameEN : "")
                                },
                                universityId: (dr.plcUniversityId ? dr.plcUniversityId : ""),
                                universityName: {
                                    TH: "",
                                    EN: (dr.universityNameEN ? dr.universityNameEN : "")
                                },
                                universityOther: (dr.universityOther ? dr.universityOther : ""),
                                facultyId: (dr.acaFacultyId ? dr.acaFacultyId : ""),
                                facultyName: {
                                    TH: "",
                                    EN: (dr.facultyNameEN ? dr.facultyNameEN : "")
                                },
                                department: (dr.department ? dr.department : ""),
                                semesterDateStart: (dr.semesterDateStart ? dr.semesterDateStart : ""),
                                semesterDateEnd: (dr.semesterDateEnd ? dr.semesterDateEnd : ""),
                                homeAddress: (dr.homeAddress ? dr.homeAddress : ""),
                                homeDistrict: (dr.homeDistrict ? dr.homeDistrict : ""),
                                homeCity: (dr.homeCity ? dr.homeCity : ""),
                                homeProvince: (dr.homeProvince ? dr.homeProvince : ""),
                                homePostalCode: (dr.homePostalCode ? dr.homePostalCode : ""),
                                countryHomeId: (dr.plcCountryHomeId ? dr.plcCountryHomeId : ""),
                                countryHomeName: {
                                    TH: "",
                                    EN: (dr.countryHomeNameEN ? dr.countryHomeNameEN : "")
                                },
                                mailingAddress: (dr.mailingAddress ? dr.mailingAddress : ""),
                                mailingDistrict: (dr.mailingDistrict ? dr.mailingDistrict : ""),
                                mailingCity: (dr.mailingCity ? dr.mailingCity : ""),
                                mailingProvince: (dr.mailingProvince ? dr.mailingProvince : ""),
                                mailingPostalCode: (dr.mailingPostalCode ? dr.mailingPostalCode : ""),
                                countryMailingId: (dr.plcCountryMailingId ? dr.plcCountryMailingId : ""),
                                countryMailingName: {
                                    TH: "",
                                    EN: (dr.countryMailingNameEN ? dr.countryMailingNameEN : "")
                                },
                                contactTel: (dr.contactTel ? dr.contactTel : ""),
                                birthDate: (dr.birthDate ? dr.birthDate : ""),
                                maritalStatusId: (dr.perMaritalStatusId ? dr.perMaritalStatusId : ""),
                                maritalStatusName: {
                                    TH: "",
                                    EN: (dr.maritalStatusNameEN ? dr.maritalStatusNameEN : "")
                                },
                                passportNumber: (dr.passportNumber ? dr.passportNumber : ""),
                                passportExpiryDate: (dr.passportExpiryDate ? dr.passportExpiryDate : ""),
                                emergencyName: (dr.emergencyName ? dr.emergencyName : ""),
                                relationshipId: (dr.perRelationshipId ? dr.perRelationshipId : ""),
                                relationshipName: {
                                    TH: "",
                                    EN: (dr.relationshipNameEN ? dr.relationshipNameEN : "")
                                },
                                emergencyMailingAddress: (dr.emergencyMailingAddress ? dr.emergencyMailingAddress : ""),
                                emergencyMailingDistrict: (dr.emergencyMailingDistrict ? dr.emergencyMailingDistrict : ""),
                                emergencyMailingCity: (dr.emergencyMailingCity ? dr.emergencyMailingCity : ""),
                                emergencyMailingProvince: (dr.emergencyMailingProvince ? dr.emergencyMailingProvince : ""),
                                emergencyMailingPostalCode: (dr.emergencyMailingPostalCode ? dr.emergencyMailingPostalCode : ""),
                                countryMailingEmergencyId: (dr.plcCountryMailingEmergencyId ? dr.plcCountryMailingEmergencyId : ""),
                                countryMailingEmergencyName: {
                                    TH: "",
                                    EN: (dr.countryMailingEmergencyNameEN ? dr.countryMailingEmergencyNameEN : "")
                                },
                                emergencyHomeTel: (dr.emergencyHomeTel ? dr.emergencyHomeTel : ""),
                                emergencyOfficeTel: (dr.emergencyOfficeTel ? dr.emergencyOfficeTel : ""),
                                emergencyMobile: (dr.emergencyMobile ? dr.emergencyMobile : ""),
                                emergencyFax: (dr.emergencyFax ? dr.emergencyFax : ""),
                                emergencyEmail: (dr.emergencyEmail ? dr.emergencyEmail : ""),
                                curEduInstitution: (dr.curEduInstitution ? dr.curEduInstitution : ""),
                                curEduFaculty: (dr.curEduFaculty ? dr.curEduFaculty : ""),
                                disciplineId: (dr.sexDisciplineId ? dr.sexDisciplineId : ""),
                                disciplineName: {
                                    TH: "",
                                    EN: (dr.disciplineNameEN ? dr.disciplineNameEN : "")
                                },
                                educationalBackgroundId: (dr.perEducationalBackgroundId ? dr.perEducationalBackgroundId : ""),
                                educationalBackgroundName: {
                                    TH: "",
                                    EN: (dr.educationalBackgroundNameEN ? dr.educationalBackgroundNameEN : "")
                                },
                                curEduYear: (dr.curEduYear ? dr.curEduYear : ""),
                                TOEFLScore: (dr.TOEFLScore ? dr.TOEFLScore : ""),
                                TOEFLTypeId: (dr.sexTOEFLTypeId ? dr.sexTOEFLTypeId : ""),
                                TOEFLTypeName: {
                                    TH: "",
                                    EN: (dr.TOEFLTypeNameEN ? dr.TOEFLTypeNameEN : "")
                                },
                                IELTSResult: (dr.IELTSResult ? dr.IELTSResult : ""),
                                englishNativeOther: (dr.englishNativeOther ? dr.englishNativeOther : ""),
                                passportFile: (dr.passportFile ? dr.passportFile : ""),
                                facePhotographsFile: (dr.facePhotographsFile ? dr.facePhotographsFile : ""),
                                transcriptFile: (dr.transcriptFile ? dr.transcriptFile : ""),
                                travelInsuranceFile: (dr.travelInsuranceFile ? dr.travelInsuranceFile : ""),
                                letterHomeUniversityFile: (dr.letterHomeUniversityFile ? dr.letterHomeUniversityFile : ""),
                                curriculumVitaeFile: (dr.curriculumVitaeFile ? dr.curriculumVitaeFile : ""),
                                TOEFLIELTSFile: (dr.TOEFLIELTSFile ? dr.TOEFLIELTSFile : ""),
                                letterIntentFile: (dr.letterIntentFile ? dr.letterIntentFile : ""),
                                declarationA: (dr.declarationA ? dr.declarationA : ""),
                                declarationAName: (dr.declarationAName ? dr.declarationAName : ""),
                                declarationB: (dr.declarationB ? dr.declarationB : ""),
                                declarationC: (dr.declarationC ? dr.declarationC : ""),
                                declarationD: (dr.declarationD ? dr.declarationD : ""),
                                declarationE: (dr.declarationE ? dr.declarationE : ""),
                                declarationEName: (dr.declarationEName ? dr.declarationEName : ""),
                                declarationESpecify: (dr.declarationESpecify ? dr.declarationESpecify : ""),
                                submitStatus: (dr.submitStatus ? dr.submitStatus : "")
                            });
                        }

                        param.dataSource = dt;

                        deferred.resolve(param.dataSource);
                    }
                    else
					    utilServ.dialogError(utilServ.getLabel(["application", "applicationNotFound"]), function (e) {
						    param.dataSource = [];

						    deferred.resolve(param.dataSource);
					    });
                });
            }
            else
                deferred.resolve(param.dataSource);

            return deferred.promise;
        };

        self.saveChange = {
            update: {
                action: function (param) {
                    param.data = (param.data === undefined || param.data === "" ? {} : param.data);

                    var deferred = $q.defer();
                    var data = param.data;

                    appServ.getCookie().then(function (result) {
                        if (result.status === true) {                                                
                            appServ.getCookieValue(["userId", "verifyCode", "applicationId", "applicationStatus"]);

                            var cookie = utilServ.getCookie({ cookieName: appServ.cookieName });                                                                          

                            angular.extend(data, {
                                "package": ((cookie !== null && cookie.key !== undefined && cookie.key.length > 0) ? appServ.setPackage([cookie.key, appServ.cookieInfo.applicationId, appServ.cookieInfo.userId, appServ.cookieInfo.verifyCode]) : ""),
                                "by": appServ.cookieInfo.userId
                            });
                        }

                        appServ.save.action({
                            url: (utilServ.getURLAPI(appServ.pathAPI) + "Application/PutData"),
                            method: "PUT",
                            data: [data]
                        }).then(function (result) {
                            deferred.resolve(result);
                        });                    
                    });

                    return deferred.promise;
                }
            }
        };

        self.submit = {            
            getValue: function () {
                var result = {
		            "submitStatus": "Y",
		            "applicationStatus": "PA"
                };

                return result;
            },
            validate: function () {
                var deferred = $q.defer();
                var dataRow = {};

                self.getDataSource({}).then(function (result) {
                    var res = {
					    status: true
                    };

                    dataRow = result[0];

                    if (res.status &&
                        (
                            (!dataRow.studentCategoryId && !dataRow.studentCategorySpecify && !dataRow.studentCategoryOther) ||
                            !dataRow.countryHomeUniversityId ||
                            (!dataRow.universityId && !dataRow.universityOther) ||
                            !dataRow.facultyId ||
                            !dataRow.semesterDateStart ||
                            !dataRow.semesterDateEnd
                        )) {
                        res.status = false;
                        res.error = {
                            msg: appServ.getLabel(["application", "studentInfo", "submit"])
                        };
                        res.location = pageRouteServ.pageObject.studentInfo.url;
                    }

                    if (res.status &&
                        (
                            !dataRow.titlePrefixId ||
                            !dataRow.firstName ||
                            !dataRow.lastName ||
                            !dataRow.birthDate ||
                            !dataRow.genderId ||
                            !dataRow.nationalityId ||
                            !dataRow.maritalStatusId ||
                            !dataRow.passportNumber ||
                            !dataRow.passportExpiryDate
                        )) {
		                res.status = false;
					    res.error = {
						    msg: appServ.getLabel(["application", "personalInfo", "submit"])
					    };
					    res.location = pageRouteServ.pageObject.personalInfo.url;
                    }

                    if (res.status &&
                        (
                            !dataRow.homeAddress ||
                            !dataRow.homeCity ||
                            !dataRow.homeProvince ||
                            !dataRow.countryHomeId
                        )) {
                        res.status = false;
                        res.error = {
						    msg: appServ.getLabel(["application", "personalInfo", "submit"])
                        };
                        res.location = pageRouteServ.pageObject.personalInfo.url;
                    }

                    if (res.status &&
                        (
                            !dataRow.mailingAddress ||
                            !dataRow.mailingCity ||
                            !dataRow.mailingProvince ||
                            !dataRow.countryMailingId
                        )) {
					    res.status = false;
					    res.error = {
						    msg: appServ.getLabel(["application", "personalInfo", "submit"])
					    };
					    res.location = pageRouteServ.pageObject.personalInfo.url;
                    }

                    if (res.status &&
                        (
                            !dataRow.email ||
                            (dataRow.email && !utilServ.isEmail(dataRow.email))
                        )) {
					    res.status = false;
					    res.error = {
						    msg: appServ.getLabel(["application", "personalInfo", "submit"])
					    };
					    res.location = pageRouteServ.pageObject.personalInfo.url;
                    }

                    if (res.status &&
                        (
                            !dataRow.emergencyName ||
                            !dataRow.relationshipId ||
                            !dataRow.emergencyMailingAddress ||
                            !dataRow.emergencyMailingCity ||
                            !dataRow.emergencyMailingProvince ||
                            !dataRow.countryMailingEmergencyId ||
                            !dataRow.emergencyHomeTel ||
                            !dataRow.emergencyMobile ||
                            !dataRow.emergencyEmail ||
                            (dataRow.emergencyEmail && !utilServ.isEmail(dataRow.emergencyEmail))
                        )) {
					    res.status = false;
					    res.error = {
						    msg: appServ.getLabel(["application", "personalInfo", "submit"])
					    };
					    res.location = pageRouteServ.pageObject.personalInfo.url;
                    }

                    if (res.status &&
                        (
                            !dataRow.curEduInstitution ||
                            !dataRow.curEduFaculty ||
                            !dataRow.disciplineId ||
                            !dataRow.educationalBackgroundId ||
                            !dataRow.curEduYear
                        )) {
					    res.status = false;
					    res.error = {
						    msg: appServ.getLabel(["application", "academicInfo", "submit"])
					    };
					    res.location = pageRouteServ.pageObject.academicInfo.url;
                    }

                    if (res.status &&
                        (
                            !dataRow.passportFile ||
                            !dataRow.facePhotographsFile ||
                            !dataRow.travelInsuranceFile
                        )) {
                        res.status = false;
                        res.error = {
						    msg: appServ.getLabel(["application", "uploadDocuments", "submit"])
                        };
                        res.location = pageRouteServ.pageObject.uploadDocuments.url;
                    }

                    if (res.status &&
                        (
                            !dataRow.declarationA ||
                            (dataRow.declarationA === "Y" && !dataRow.declarationB) ||
                            !dataRow.declarationC ||
                            !dataRow.declarationD ||
                            !dataRow.declarationE ||
                            (dataRow.declarationE === "Y" && !dataRow.declarationESpecify)
                        )) {
                        res.status = false;
                        res.error = {
	                        msg: appServ.getLabel(["application", "declaration", "submit"])
                        };
                        res.location = pageRouteServ.pageObject.declaration.url;
                    }

                    $timeout(function () {
                        deferred.resolve(res);
                    }, 0);
                });

                return deferred.promise;
            },
            action: function () {
                utilServ.getDialogPreloading(dictServ.dict.msgPreloading.submitting);

                this.validate().then(function (result) {
                    if (result.status) {
                        self.saveChange.update.action({
                            data: self.submit.getValue()
                        }).then(function (result) {
                            appServ.closeDialogPreloading();

                            self.isSubmitStatus = result.status;

                            if (result.status) {                                                     
                                $rootScope.$appServ.getUserInfo.applicationStatus = "PA";
                                $rootScope.$appServ.getUserInfo.applicationStatusName = "Pending Approval";
                                
                                pageRouteServ.setSlideMenuOff();

                                appServ.getCookieValue(["userId", "verifyCode", "applicationId", "applicationStatus"]);

                                appServ.setCookie({
                                    userId: appServ.cookieInfo.userId,
                                    verifyCode: appServ.cookieInfo.verifyCode,
                                    applicationId: appServ.cookieInfo.applicationId,
                                    applicationStatus: $rootScope.$appServ.getUserInfo.applicationStatus
                                });
                            }
                        });
                    }
                    else {
                        appServ.closeDialogPreloading();

                        utilServ.dialogError(result.error.msg, function (e) {
						    if (e === "close")
							    $timeout(function () {
						            $location.path(result.location).replace();
							    }, 0);
                        });
                    }                    
                });
            }
        };
    })        

    .run(function ($rootScope, applicationServ) {
        $rootScope.$applicationServ = applicationServ;
    });
})();

File: API/Models/Application.cs
/*
=============================================
Author      : <ยุทธภูมิ ตวันนา>
Create date : <๒๗/๐๖/๒๕๖๑>
Modify date : <๐๘/๐๔/๒๕๖๕>
Description : <โมเดลข้อมูลใบสมัคร>
=============================================
*/

using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;

namespace API.Models {
	public class Application {
		public string package {
			get;
			set;
		}

		public string titlePrefixId {
			get;
			set;
		}

		public string firstName {
			get;
			set;
		}

		public string middleName {
			get;
			set;
		}

		public string lastName {
			get;
			set;
		}

		public string nationalityId {
			get;
			set;
		}

		public string genderId {
			get;
			set;
		}

		public string email {
			get;
			set;
		}

		public string studentCategoryId {
			get;
			set;
		}

		public string studentCategorySpecify {
			get;
			set;
		}

		public string studentCategoryOther {
			get;
			set;
		}

		public string countryId {
			get;
			set;
		}

		public string universityId {
			get;
			set;
		}

		public string universityOther {
			get;
			set;
		}

		public string facultyId {
			get;
			set;
		}

		public string department {
			get;
			set;
		}

		public string semesterDateStart {
			get;
			set;
		}

		public string semesterDateEnd {
			get;
			set;
		}

		public string homeAddress {
			get;
			set;
		}

		public string homeDistrict {
			get;
			set;
		}

		public string homeCity {
			get;
			set;
		}

		public string homeProvince {
			get;
			set;
		}

		public string homePostalCode {
			get;
			set;
		}

		public string countryHomeId {
			get;
			set;
		}

		public string mailingAddress {
			get;
			set;
		}

		public string mailingDistrict {
			get;
			set;
		}

		public string mailingCity {
			get;
			set;
		}

		public string mailingProvince {
			get;
			set;
		}

		public string mailingPostalCode {
			get;
			set;
		}

		public string countryMailingId {
			get;
			set;
		}

		public string contactTel {
			get;
			set;
		}

		public string birthdate {
			get;
			set;
		}

		public string maritalStatusId {
			get;
			set;
		}

		public string passportNumber {
			get;
			set;
		}

		public string passportExpiryDate {
			get;
			set;
		}

		public string emergencyName {
			get;
			set;
		}

		public string relationshipId {
			get;
			set;
		}

		public string emergencyMailingAddress {
			get;
			set;
		}

		public string emergencyMailingDistrict {
			get;
			set;
		}

		public string emergencyMailingCity {
			get;
			set;
		}

		public string emergencyMailingProvince {
			get;
			set;
		}

		public string emergencyMailingPostalCode {
			get;
			set;
		}

		public string countryMailingEmergencyId {
			get;
			set;
		}

		public string emergencyHomeTel {
			get;
			set;
		}

		public string emergencyOfficeTel {
			get;
			set;
		}

		public string emergencyMobile {
			get;
			set;
		}

		public string emergencyFax {
			get;
			set;
		}

		public string emergencyEmail {
			get;
			set;
		}

		public string curEduInstitution {
			get;
			set;
		}

		public string curEduFaculty {
			get;
			set;
		}

		public string disciplineId {
			get;
			set;
		}

		public string educationalBackgroundId {
			get;
			set;
		}

		public string curEduYear {
			get;
			set;
		}

		public string TOEFLScore {
			get;
			set;
		}

		public string TOEFLTypeId {
			get;
			set;
		}

		public string IELTSResult {
			get;
			set;
		}

		public string englishNativeOther {
			get;
			set;
		}

		public string passportFile {
			get;
			set;
		}

		public string facePhotographsFile {
			get;
			set;
		}

		public string transcriptFile {
			get;
			set;
		}

		public string travelInsuranceFile {
			get;
			set;
		}

		public string letterHomeUniversityFile {
			get;
			set;
		}

		public string curriculumVitaeFile {
			get;
			set;
		}

		public string TOEFLIELTSFile {
			get;
			set;
		}

		public string letterIntentFile {
			get;
			set;
		}

		public string declarationA {
			get;
			set;
		}

		public string declarationB {
			get;
			set;
		}

		public string declarationC {
			get;
			set;
		}

		public string declarationD {
			get;
			set;
		}

		public string declarationE {
			get;
			set;
		}

		public string declarationESpecify {
			get;
			set;
		}

		public string submitStatus {
			get;
			set;
		}

		public string submitBy {
			get;
			set;
		}

		public string submitIp {
			get;
			set;
		}

		public string applicationStatus {
			get;
			set;
		}

		public string centralApprovalStatus {
			get;
			set;
		}

		public string centralApprovalBy {
			get;
			set;
		}

		public string centralApprovalIp {
			get;
			set;
		}

		public string divisionApprovalStatus {
			get;
			set;
		}

		public string divisionApprovalBy {
			get;
			set;
		}

		public string divisionApprovalIp {
			get;
			set;
		}

		public string reason {
			get;
			set;
		}

		public string cancelledStatus {
			get;
			set;
		}

		public string by {
			get;
			set;
		}

		public static DataSet GetData(
			string applicationId,
			string userId,
			string verifyCode
		) {
			DataSet ds = iUtil.ExecuteCommandStoredProcedure(iUtil.bermudaConnectionString, "sp_sexGetApplication",
				new SqlParameter("@id", applicationId),
				new SqlParameter("@userId", userId),
				new SqlParameter("@verifyCode",	verifyCode)
			);

			return ds;
		}

		public static DataSet SetData(
			string method,
			List<Application> data
		) {
			string action = String.Empty;
			string cookieValue = String.Empty;
			string[] packageDecode = null;
			string applicationId = String.Empty;
			string userId = String.Empty;
			string verifyCode = String.Empty;
			DataSet ds = new DataSet();
			StudentService.StudentService ss = new StudentService.StudentService();

			if (method.Equals("POST"))
				action = "INSERT";

			if (method.Equals("PUT"))
				action = "UPDATE";

			if (method.Equals("DELETE"))
				action = "DELETE";
            
			foreach (var d in data) {
				try { 
					packageDecode = ss.DecodeBase64String(d.package).Split('.');

					if (action.Equals("INSERT")) {
						userId = ss.DecodeBase64String(ss.StringReverse(packageDecode[0]));
						verifyCode = ss.DecodeBase64String(ss.StringReverse(packageDecode[1]));
					}

					if (action.Equals("UPDATE")) {
						cookieValue = ss.DecodeBase64String(ss.StringReverse(packageDecode[0]));

						if (iUtil.CompareCookie(iUtil.cookieName, cookieValue)) {
							applicationId = ss.DecodeBase64String(ss.StringReverse(packageDecode[1]));
							userId = ss.DecodeBase64String(ss.StringReverse(packageDecode[2]));
							verifyCode = ss.DecodeBase64String(ss.StringReverse(packageDecode[3]));
						}
					}

					ds = iUtil.ExecuteCommandStoredProcedure(iUtil.bermudaConnectionString, "sp_sexSetApplication",
						new SqlParameter("@action", action),
						new SqlParameter("@id", applicationId),
						new SqlParameter("@userId", userId),
						new SqlParameter("@verifyCode", verifyCode),
						new SqlParameter("@titlePrefixId", d.titlePrefixId),
						new SqlParameter("@firstName", d.firstName),
						new SqlParameter("@middleName", d.middleName),
						new SqlParameter("@lastName", d.lastName),
						new SqlParameter("@nationalityId", d.nationalityId),
						new SqlParameter("@genderId", d.genderId),
						new SqlParameter("@email", d.email),
						new SqlParameter("@studentCategoryId", d.studentCategoryId),
						new SqlParameter("@studentCategorySpecify", d.studentCategorySpecify),
						new SqlParameter("@studentCategoryOther", d.studentCategoryOther),
						new SqlParameter("@countryId", d.countryId),
						new SqlParameter("@universityId", d.universityId),
						new SqlParameter("@universityOther", d.universityOther),
						new SqlParameter("@facultyId", d.facultyId),
						new SqlParameter("@department", d.department),
						new SqlParameter("@semesterDateStart", d.semesterDateStart),
						new SqlParameter("@semesterDateEnd", d.semesterDateEnd),
						new SqlParameter("@homeAddress", d.homeAddress),
						new SqlParameter("@homeDistrict", d.homeDistrict),
						new SqlParameter("@homeCity", d.homeCity),
						new SqlParameter("@homeProvince", d.homeProvince),
						new SqlParameter("@homePostalCode", d.homePostalCode),
						new SqlParameter("@countryHomeId", d.countryHomeId),
						new SqlParameter("@mailingAddress", d.mailingAddress),
						new SqlParameter("@mailingDistrict", d.mailingDistrict),
						new SqlParameter("@mailingCity", d.mailingCity),
						new SqlParameter("@mailingProvince", d.mailingProvince),
						new SqlParameter("@mailingPostalCode", d.mailingPostalCode),
						new SqlParameter("@countryMailingId", d.countryMailingId),
						new SqlParameter("@contactTel", d.contactTel),
						new SqlParameter("@birthdate", d.birthdate),
						new SqlParameter("@maritalStatusId", d.maritalStatusId),
						new SqlParameter("@passportNumber", d.passportNumber),
						new SqlParameter("@passportExpiryDate", d.passportExpiryDate),
						new SqlParameter("@emergencyName", d.emergencyName),
						new SqlParameter("@relationshipId", d.relationshipId),
						new SqlParameter("@emergencyMailingAddress", d.emergencyMailingAddress),
						new SqlParameter("@emergencyMailingDistrict", d.emergencyMailingDistrict),
						new SqlParameter("@emergencyMailingCity", d.emergencyMailingCity),
						new SqlParameter("@emergencyMailingProvince", d.emergencyMailingProvince),
						new SqlParameter("@emergencyMailingPostalCode",	d.emergencyMailingPostalCode),
						new SqlParameter("@countryMailingEmergencyId", d.countryMailingEmergencyId),
						new SqlParameter("@emergencyHomeTel", d.emergencyHomeTel),
						new SqlParameter("@emergencyOfficeTel", d.emergencyOfficeTel),
						new SqlParameter("@emergencyMobile", d.emergencyMobile),
						new SqlParameter("@emergencyFax", d.emergencyFax),
						new SqlParameter("@emergencyEmail", d.emergencyEmail),
						new SqlParameter("@curEduInstitution", d.curEduInstitution),
						new SqlParameter("@curEduFaculty", d.curEduFaculty),
						new SqlParameter("@disciplineId", d.disciplineId),
						new SqlParameter("@educationalBackgroundId", d.educationalBackgroundId),
						new SqlParameter("@curEduYear", d.curEduYear),
						new SqlParameter("@TOEFLScore", d.TOEFLScore),
						new SqlParameter("@TOEFLTypeId", d.TOEFLTypeId),
						new SqlParameter("@IELTSResult", d.IELTSResult),
						new SqlParameter("@englishNativeOther", d.englishNativeOther),
						new SqlParameter("@passportFile", d.passportFile),
						new SqlParameter("@facePhotographsFile", d.facePhotographsFile),
						new SqlParameter("@transcriptFile", d.transcriptFile),
						new SqlParameter("@travelInsuranceFile", d.travelInsuranceFile),
						new SqlParameter("@letterHomeUniversityFile", d.letterHomeUniversityFile),
						new SqlParameter("@curriculumVitaeFile", d.curriculumVitaeFile),
						new SqlParameter("@TOEFLIELTSFile", d.TOEFLIELTSFile),
						new SqlParameter("@letterIntentFile", d.letterIntentFile),
						new SqlParameter("@declarationA", d.declarationA),
						new SqlParameter("@declarationB", d.declarationB),
						new SqlParameter("@declarationC", d.declarationC),
						new SqlParameter("@declarationD", d.declarationD),
						new SqlParameter("@declarationE", d.declarationE),
						new SqlParameter("@declarationESpecify", d.declarationESpecify),
						new SqlParameter("@submitStatus", d.submitStatus),
						new SqlParameter("@submitBy", d.by),
						new SqlParameter("@submitIp", iUtil.GetIP()),
						new SqlParameter("@applicationStatus", d.applicationStatus),
						new SqlParameter("@centralApprovalStatus", d.centralApprovalStatus),
						new SqlParameter("@centralApprovalBy", d.by),
						new SqlParameter("@centralApprovalIp", iUtil.GetIP()),
						new SqlParameter("@divisionApprovalStatus", d.divisionApprovalStatus),
						new SqlParameter("@divisionApprovalBy", d.by),
						new SqlParameter("@divisionApprovalIp", iUtil.GetIP()),
						new SqlParameter("@reason", d.reason),
						new SqlParameter("@cancelledStatus", d.cancelledStatus),
						new SqlParameter("@by", d.by),
						new SqlParameter("@ip", iUtil.GetIP())
					);

					if (ds.Tables[0].Rows.Count > 0) {
						DataRow dr = ds.Tables[0].Rows[0];

						if (dr["success"].ToString().Equals("Y")) {
							if (action.Equals("UPDATE") && dr["submitStatus"].ToString().Equals("Y"))
								iUtil.SendMail(dr["mailRecipients"].ToString(), dr["mailSubject"].ToString(), dr["mailMessage"].ToString());
						}

						dr["mailMessage"] = null;
					}
				}
				catch {
				}
			}

			return ds;
		}
	}
}